tests/test_traced_data.py

```diff
@@ -0,0 +1,301 @@
+import collections
+import time
+import unittest
+import six
+
+from core_data_modules.traced_data import TracedData, Metadata
+
+
+class TestTracedData(unittest.TestCase):
+    @staticmethod
+    def td_1():
+        data = {"id": "0", "phone": "+441632000001", "gender": "man"}
+        return TracedData(data, Metadata("test_user", "run_fetcher", time.time()))
+
+    @classmethod
+    def td_2(cls, td):
+        data_cleaned = {"gender": "male", "age": 30}
+        td.append(data_cleaned, Metadata("test_user", "demographic_cleaner", time.time()))
+
+    def test_append(self):
+        td = self.td_1()
+
+        self.assertEqual(td.get("id"), "0")
+        self.assertEqual(td.get("phone"), "+441632000001")
+        self.assertEqual(td.get("gender"), "man")
+        self.assertEqual(td.get("age"), None)
+        self.assertEqual(td.get("age", "default"), "default")
+
+        self.td_2(td)
+
+        self.assertEqual(td.get("id"), "0")
+        self.assertEqual(td.get("phone"), "+441632000001")
+        self.assertEqual(td.get("gender"), "male")
+        self.assertEqual(td.get("age"), 30)
+
+        # Test that the original data is still available.
+        history = td.get_history("gender")
+        self.assertListEqual(list(map(lambda x: x["value"], history)), ["man", "male"])
+
+    def test___len__(self):
+        td = self.td_1()
+        self.assertEqual(len(td), 3)
+
+        self.td_2(td)
+        self.assertEqual(len(td), 4)
+
+    def test___contains__(self):
+        td = self.td_1()
+
+        self.assertEqual("gender" in td, True)
+        self.assertEqual("age" in td, False)
+
+        self.assertEqual("gender" not in td, False)
+        self.assertEqual("age" not in td, True)
+
+        self.td_2(td)
+
+        self.assertEqual("gender" in td, True)
+        self.assertEqual("age" in td, True)
+
+    def test_items(self):
```

Comment:
These methods are getting pretty difficult to read - could we split to add new version specific methods

test_items_PY2 and test_items PY3 and have the base method do nothing other than split which gets called?

Reply:
Reorganised to match the structure of TracedData